### ISSUE TITLE
Accessing last index

### ISSUE THREAD

**Issue by Balinus** [JuliaArrays/AxisArrays.jl#60]
Hello, first of all: nice package! It's doing what I was trying to build in my own package. Now, I'm more into integrating this package into my own. :)

I do have one issue so far though: I'm building some AxisArrays with a time, longitude and latitude axis. Using the following syntax, I can't access the last element of the time dimension using `end`.

edit - using:
`Julia 0.5`
`AxisArrays 0.0.4`

```julia
julia> typeof(test)
AxisArrays.AxisArray{Float64,3,Array{Float64,3},Tuple{AxisArrays.Axis{:time,Array{Date,1}},AxisArrays.Axis{:lon,Array{Float64,1}},AxisArrays.Axis{:lat,Array{Float64,1}}}}

julia> T = test[AxisArrays.Axis{:time}]
AxisArrays.Axis{:time,Array{Date,1}}(Date[1950-01-01,1950-01-02,1950-01-03,1950-01-04,1950-01-05,1950-01-06,1950-01-07,1950-01-08,1950-01-09,1950-01-10  …  2100-12-22,2100-12-23,2100-12-24,2100-12-25,2100-12-26,2100-12-27,2100-12-28,2100-12-29,2100-12-30,2100-12-31])

julia> typeof(T)
AxisArrays.Axis{:time,Array{Date,1}}

julia> T[1]
1950-01-01

julia> T[end]
MethodError: no method matching endof(::AxisArrays.Axis{:time,Array{Date,1}})
Closest candidates are:
  endof(::SimpleVector) at essentials.jl:169
  endof(::Char) at char.jl:18
  endof(::String) at strings/string.jl:37
  ...

julia> T[55115]
2100-12-31
```

Another example:

```julia
julia> test[Axis{:time}(1)]
2-dimensional AxisArray{Float64,2,...} with axes:
    :lon, [0.0,2.8125,5.625,8.4375,11.25,14.0625,16.875,19.6875,22.5,25.3125  …  331.875,334.688,337.5,340.313,343.125,345.938,348.75,351.563,354.375,357.188]
    :lat, [-87.8638,-85.0965,-82.3129,-79.5256,-76.7369,-73.9475,-71.1578,-68.3678,-65.5776,-62.7874  …  62.7874,65.5776,68.3678,71.1578,73.9475,76.7369,79.5256,82.3129,85.0965,87.8638]
And data, a 128×64 Array{Float64,2}:
 -13.1781   -9.32368   -7.82642   -8.93107   -9.87894  …   -1.83304   -4.26288  -10.9854   -22.7088  -32.7322  -37.4845
 -13.1757   -9.49341   -8.18998   -9.56293  -10.5652       -2.10532   -3.36542   -8.54114  -20.7078  -31.7471  -37.1273
 -13.2207   -9.75245   -8.60462  -10.1924   -11.126        -2.91254   -3.39817   -7.39728  -18.8907  -30.91    -36.9068
 -13.4053  -10.0192    -9.02738  -10.6354   -11.8759       -4.08307   -4.07075   -6.77033  -17.479   -30.0259  -36.6732
 -13.533   -10.9317    -9.48826  -11.1507   -12.4858       -5.39445   -5.13422   -7.10557  -16.6137  -29.0839  -36.4522
 -13.6584  -10.7656    -9.93928  -11.7239   -12.8303   …   -6.65769   -7.51227   -8.27405  -16.1455  -28.2062  -36.322 
 -13.6874  -10.9521   -10.4106   -12.1195   -13.2295       -8.16553  -10.9555    -8.8057   -15.9809  -27.546   -36.2891
 -13.7967  -11.1882   -10.8646   -12.5337   -13.5852       -9.87189  -12.7761    -9.81238  -16.2154  -27.0488  -36.325 
 -13.8111  -11.4431   -11.3259   -13.0253   -13.9069      -11.5795   -15.156    -11.1282   -16.5414  -26.7977  -36.3877
 -13.9304  -11.7004   -11.7421   -13.4728   -14.2188      -13.0959   -16.7264   -15.3921   -16.9609  -26.8751  -36.26  
 -14.0054  -11.9974   -12.1023   -13.8859   -14.5353   …  -15.4631   -20.6427   -17.7955   -17.3621  -27.1115  -36.0593
 -14.0489  -12.3577   -12.4554   -14.2304   -14.8702      -16.9788   -23.8476   -19.3232   -18.0571  -27.1679  -35.8542
 -14.1056  -12.4714   -12.8453   -14.4906   -15.1919      -20.4497   -26.3586   -20.5891   -18.5888  -27.1167  -35.6224
 -14.2416  -12.7158   -13.2491   -14.738    -15.3265      -22.8977   -28.4267   -21.5647   -19.1263  -27.1368  -35.4454
 -14.1789  -12.9751   -13.5664   -14.798    -15.3577      -24.4087   -29.7131   -22.5045   -19.7391  -27.3171  -35.3752
 -14.3764  -13.238    -13.8794   -14.9066   -15.2137   …  -25.6306   -30.6848   -23.4341   -20.2359  -27.5961  -35.4666
   ⋮                                                   ⋱                          ⋮                                    
 -12.6153   -7.49509   -4.35917   -3.55839   -2.69089     -30.2057   -30.5568   -32.3227   -37.9244  -38.7628  -38.3909
 -12.5765   -7.47901   -4.42795   -3.67417   -3.07596     -31.7261   -34.4672   -33.0697   -38.7802  -39.4308  -38.6722
 -12.5693   -7.47721   -4.52579   -3.77119   -2.93836     -33.7065   -35.7553   -33.3518   -39.5584  -40.0128  -38.9905
 -12.6683   -7.48649   -4.58289   -3.85673   -2.86304  …  -35.142    -35.927    -33.8506   -40.2003  -40.5507  -39.2897
 -12.7379   -7.51856   -4.69724   -3.96305   -2.78657     -32.0892   -37.2641   -34.1958   -40.5407  -41.0779  -39.5395
 -12.6542   -7.57889   -4.86066   -4.07657   -2.7201      -32.0626   -36.9486   -34.3959   -40.6646  -41.5107  -39.7349
 -12.643    -7.64567   -4.9675    -4.24183   -2.78983     -27.3247   -35.6354   -34.0476   -40.3867  -41.3766  -39.8855
 -12.7144   -7.73249   -5.1489    -4.43427   -2.97227     -25.3403   -33.3752   -33.4608   -39.6832  -40.7994  -40.0125
 -12.7472   -7.83204   -5.366     -4.72288   -3.25089  …  -21.6124   -29.8843   -33.0288   -39.0629  -40.2828  -39.9988
 -12.7618   -7.94227   -5.5405    -5.03351   -3.53205     -16.5938   -23.5874   -31.3431   -37.6958  -39.6077  -39.7354
 -12.772    -8.08826   -5.77714   -5.37885   -4.11234     -13.8241   -20.7749   -26.1037   -36.3062  -38.7978  -39.4562
 -12.8576   -8.25623   -6.13535   -5.85761   -4.90565     -10.706    -18.2744   -24.2676   -33.6903  -37.878   -39.1897
 -12.9316   -8.43967   -6.46885   -6.31055   -5.83805      -8.03733  -15.184    -22.3596   -30.9707  -36.882   -39.0329
 -12.9815   -8.64146   -6.78547   -6.95096   -6.82795  …   -4.94059  -11.9601   -19.3953   -29.2095  -35.781   -38.3263
 -13.0144   -8.85602   -7.04481   -7.52024   -7.87766      -3.31864   -9.8346   -16.7155   -27.2396  -34.8338  -38.1462
 -13.0999   -9.158     -7.44782   -8.31867   -8.89557      -2.30311   -6.4213   -13.8705   -24.8114  -33.7981  -37.8232

```

```julia
julia> test[Axis{:time}(end)]
------ BoundsError --------------------- Stacktrace (most recent call last)

 [1] — getindex(::AxisArrays.AxisArray{Float64,3,Array{Float64,3},Tuple{AxisArrays.Axis{:time,Array{Date,1}},AxisArrays.Axis{:lon,Array{Float64,1}},AxisArrays.Axis{:lat,Array{Float64,1}}}}, ::AxisArrays.Axis{:time,Int64}) at indexing.jl:84

 [2] — getindex at indexing.jl:60 [inlined]

 [3] — getindex at abstractarray.jl:752 [inlined]

 [4] — _getindex at multidimensional.jl:270 [inlined]

 [5] — checkbounds at abstractarray.jl:284 [inlined]

 [6] — throw_boundserror(::Array{Float64,3}, ::Tuple{Int64,Colon,Colon}) at abstractarray.jl:355

BoundsError: attempt to access 55115×128×64 Array{Float64,3} at index [451502080,Colon(),Colon()]
```

**Comment by mbauman** [comment]
Thanks for the report!  These are two separate issues; the first is simply that the `Axis` type is missing the `endof` method.  That's a quick and easy fix.  If you're interested in contributing (and beat me to it) a pull request would be very welcome.

Accessing `test[Axis{:time}(end)]`, however, is different.  The `end` keyword is expanded at parse-time based purely upon two things: the object it is indexing into (that is, the nearest set of square brackets that contains it), and its order in that indexing expression.  It does not know that it gets wrapped by an Axis or the manner in which it might get re-ordered.  It only knows that it's the last accessible index along the first dimension… and since there are no other indices following it, it is the last _linear_ index.  You can see how this works with `expand` since it's purely a syntactic transformation:

```julia
julia> expand(:(A[end]))
:(getindex(A, (Base.endof)(A)))

julia> expand(:(A[end, end]))
:(getindex(A, (Base.size)(A, 1), (Base.trailingsize)(A, 2)))

julia> expand(:(A[end, end, end]))
:(getindex(A, (Base.size)(A, 1), (Base.size)(A, 2), (Base.trailingsize)(A, 3)))
```

The exact same thing happens when `end` gets wrapped in a specific Axis:

```julia
julia> expand(:(A[Axis{:col}(end), Axis{:row}(end)]))
:(getindex(A, ((Core.apply_type)(Axis, :col))((Base.size)(A, 1)), ((Core.apply_type)(Axis, :row))((Base.trailingsize)(A, 2))))
```

So even though it looks like I'm asking for the last column, it's actually passing `size(A, 1)` to that Axis.  We should probably add a warning in the documentation about this gotcha.

**Comment by Balinus** [comment]
Thanks for the answer!

I did a [PR](https://github.com/JuliaArrays/AxisArrays.jl/pull/61). Hopefully, the fix is good :)

**Comment by timholy** [comment]
Fixed by #61.
